libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,259 @@
+		fileLock:  flock.New(lockfilePath),
+		logger:    logp.L(),
+		beatName:  beatInfo.Beat,
+		filePath:  lockfilePath,
+		beatStart: beatInfo.StartTime,
+	}
+}
+
+// Lock attempts to acquire a lock on the data path for the currently-running
+// Beat instance. If another Beats instance already has a lock on the same data path
+// an ErrAlreadyLocked error is returned.
+func (lock *Locker) Lock() error {
+	new := pidfile{Pid: pidFetch(), WriteTime: time.Now()}
+	encoded, err := json.Marshal(&new)
+	if err != nil {
+		return fmt.Errorf("error encoding json for pidfile: %w", err)
+	}
+
+	// The combination of O_CREATE and O_EXCL will ensure we return an error if we don't
+	// manage to create the file
+	fh, openErr := os.OpenFile(lock.filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o600)
+	// Don't trust different OSes to report the errors we expect, just try to recover regardless
+	if openErr != nil {
+		err = lock.handleFailedCreate()
+		if err != nil {
+			return fmt.Errorf("cannot obtain lockfile: %w", err)
+		}
+		// If something fails here, it's probably unrecoverable
+		fh, err = os.OpenFile(lock.filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o600)
+		if err != nil {
+			return fmt.Errorf("cannot re-obtain lockfile %s: %w", lock.filePath, err)
+		}
+	}
+
+	// a Process can't write to its own locked file on all platforms, write first
+	_, err = fh.Write(encoded)
+	if err != nil {
+		return fmt.Errorf("error writing pidfile to %s: %w", lock.filePath, err)
+	}
+
+	// Exclusive lock
+	isLocked, err := lock.fileLock.TryLock()
+	if err != nil {
+		return fmt.Errorf("unable to lock data path: %w", err)
+	}
+	// case: lock could not be obtained.
+	if !isLocked {
+		// if we're here, things are probably unrecoverable, as we've previously checked for a lockfile. Exit.
+		return ErrAlreadyLocked
+	}
+
+	return nil
+}
+
+// Unlock attempts to release the lock on a data path previously acquired via Lock().
+func (lock *Locker) Unlock() error {
+	err := lock.fileLock.Unlock()
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+
+	err = os.Remove(lock.fileLock.Path())
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path file %s: %w", lock.fileLock.Path(), err)
+	}
+	return nil
+}
+
+// ******* private helpers
+
+// handleFailedCreate will attempt to recover from a failed lock operation in a pid-aware way.
+// The point of this is to deal with instances where an improper beat shutdown left us with
+// a pre-existing pidfile for a beat process that no longer exists.
+// The one argument tells the method if a pre-existing lockfile was already found.
+func (lock *Locker) handleFailedCreate() error {
+	// Try to lock the file.
+	// Not all OSes will fail on this.
+	_, err := lock.fileLock.TryLock()
+	// Case: the file already locked, and in use by another process.
+	if err != nil {
+		if runtime.GOOS == "windows" {
+			// on windows, locks from dead PIDs will be auto-released, but it might take the OS a while.
```

Comment:
I think it would be good to explicitly mention that we don't know if 1 second is sufficient, but we're giving Windows a short grace period. Perhaps quote the [`LockFile` docs](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfile):

> However, the time it takes for the operating system to unlock these locks depends upon available system resources.